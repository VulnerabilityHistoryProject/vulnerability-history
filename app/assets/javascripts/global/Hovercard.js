import $ from "jquery";

/**
 * Set a hovercard to load and present the selected vulnerability.
 * @param {*} elementSelect The class name of the elements that will have a hovercard.
 */
export default function setHovercard(elementSelect) {
  // Force cve links to stay on same line
  $(elementSelect).css("white-space", "nowrap");
  // Add hovercard div to elements with the given selector
  $(elementSelect).append('<div class="hover-card" id="unpopulated"></div>');

  // Initiate the hovercard element style and hide it
  $(".hover-card")
    .css({
      "white-space": "normal", // hovercard contents wrap
      display: "inline-flex",
      position: "absolute",
      backgroundColor: "#23232c",
      borderRadius: "15pt",
      borderLeft: "4pt solid #98e3b7",
      padding: "1.2em",
      height: "fit-content",
      width: "fit-content",
      fontSize: "12pt", //TODO fixed size or relative?
      color: "#98e3b7",
    })
    .hide();

  // Show the hovercard when hovering over the given elements and hide it when
  // hovering off it and the hovercard
  $(elementSelect).hover(
    function (e) {
      // when hovering over a link
      if ($(this).find(".hover-card").attr("id") === "unpopulated") {
        // if the card is not already populated make ajax request and change to populated
        ajaxReq($(this));
        $(this).find(".hover-card").attr("id", "populated");
      }
      $(this)
        .find(".hover-card")
        .delay(450) // wait 450ms before displaying card
        .css({
          // spawn the hovercard centered at the cursor right below the chosen element
          left: e.pageX - 100, // center with the cursor //TODO fixed size or relative?
          top: $(this)[0].getBoundingClientRect().bottom + window.scrollY - 3, // right below the element
        })
        .fadeIn(100); // make card visible
    },
    function () {
      // when hovering off the link
      $(this)
        .find(".hover-card")
        .stop(true) // if hover off before delay timer don't display card at all
        .fadeOut(100); // fade card away
    },
  );
}

// Populate hover card from given result json
function populateHovercard(result, element, cve) {
  const monthNames = [
    "Jan",
    "Feb",
    "Mar",
    "Apr",
    "May",
    "Jun",
    "Jul",
    "Aug",
    "Sep",
    "Oct",
    "Nov",
    "Dec",
  ];
  let nicknameText;
  const titleText =
    result["nickname"] === ""
      ? cve // if there is no nickname the title is the cve
      : (() => {
          //if there is a nickname it becomes the title and the cve becomes the nickname
          nicknameText = cve;
          return result["nickname"];
        })();
  const title = $("<div></div>").text(titleText).css({
    display: "flex",
    flexDirection: "column",
    fontWeight: "bold",
    textAlign: "left",
  });
  const title_description_div = $("<div></div>").css({
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
    gap: "1em",
  });
  const nickname = $("<span></span>").text(nicknameText).css({
    fontSize: "0.8em",
    fontWeight: "100",
    fontStyle: "italic",
  });
  // TODO Add the date of last event
  // let date = new Date(result["announced"].substring(0, 10));
  // const announced =
  //   monthNames[date.getMonth()] +
  //   " " +
  //   date.getDate() +
  //   ", " +
  //   date.getFullYear();
  // const announced_date = $("<span></span>").text(announced).css({
  //   fontSize: "0.6em",
  //   fontWeight: "650",
  //   color: "#ea5f49",
  // });
  // TODO Fetch and insert tag elements
  const tags_array = [
    ...new Set(result["tags_array"].slice(2, -2).split('","')),
  ].reverse();
  const tags = $("<div></div>").css({
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    justifyContent: "space-between",
    fontSize: "1.6em",
    paddingRight: "1em",
  });
  tags_array.length = 6;
  $.each(tags_array, function (index, tag) {
    tag = tag.slice(1, -1).split(",");
    tags.append(`<i class="vhp-icon-${tag[0]}" style="color: ${tag[1]}"></i>`);
  });
  const description = $("<p></p>")
    .text(truncateWords(result["description"], 40))
    .css({
      fontSize: "0.8em",
      textAlign: "left",
      color: "white",
      lineHeight: "1.5em",
      maxWidth: "27ch",
    });
  // title.append(nickname, announced_date);
  title.append(nickname);
  title_description_div.append(title, description);
  $(element).find(".hover-card").append(tags, title_description_div);
}

// fetch request for vulnerability data
function ajaxReq(element) {
  const regexp = /(CVE-\d{4}-\d{4,})/gi;
  const cve = element.text().match(regexp)[0];
  $.ajax({
    url: "/api/vulnerabilities/gettags/" + cve,
    success: (response) => {
      populateHovercard(response[0], element, cve);
    },
  });
}

// Truncate given text to a given number of words
function truncateWords(text, maxWords) {
  const words = text.split(" ");
  if (words.length > maxWords) {
    return words.slice(0, maxWords).join(" ") + "...";
  }
  return text;
}
