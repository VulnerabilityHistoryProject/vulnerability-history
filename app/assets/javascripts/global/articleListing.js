// import $ from 'jquery'
// import _ from 'lodash'
import {vhpMarkdown} from './vhpMarkdown'

export default function articleListing(jsonData, container = '#article-listing') {
  const blurbCounts = {
    '#blurb-1big-template' : 1,
    '#blurb-4even-template' : 4,
    '#blurb-1wide-template' : 1,
    '#blurb-1big1small-template' : 2,
  }
  const numTemplates = _.size(blurbCounts);
  const bMax = _.reduce(blurbCounts, (max, v, _k) => max = v > max ? v : max, 0)
  let articlesLeft = jsonData.length;
  let i = 0;
  while(articlesLeft > bMax){
    const templateID = Object.keys(blurbCounts)[i];
    addBlurbRow(container, templateID)
    i = (i + 1) % numTemplates;
    articlesLeft -= blurbCounts[templateID];
  }
  //for the last few we add a single wide row.
  _.times(articlesLeft, () => addBlurbRow(container, '#blurb-1wide-template'))

  $('.hidden').remove() //Remove leftover blurb templates

  let blurbs = $('.blurb-text')
  let blurbArts = $('.blurb-art')
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i]
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`)
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
`) + `<a href="/articles/${article.slug}" class="button read-more">Read more</a>`
  }
}

export function articleListingOneSize(jsonData, container = '#article-listing-same') {
  _.times(jsonData.length,() => addBlurbRow(container, '#blurb-1wide-template'))
  let blurbs = $(container + ' .blurb-text');
  let blurbArts = $(container + ' .blurb-art');
  for (let i = 0; i < jsonData.length; i++) {
    let article = jsonData[i];
    $(blurbArts[i]).html(`<div class="vhp-art-${article.art}"/>`);
    blurbs[i].innerHTML = vhpMarkdown(`
### ${article.title} ###
${article.blurb}
[Full Article](/articles/${article.slug})
    `);
  }
}

function addBlurbRow(container, templateID) {
  const blurbRow = $(templateID).clone();
  blurbRow.removeAttr('id').removeClass('hidden');
  $(container).append(blurbRow).append(`<hr/>`);
}
