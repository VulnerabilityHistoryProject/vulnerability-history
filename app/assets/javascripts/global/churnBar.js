export function churnBar(insertions, deletions) {
  const i_blocks = Math.log10(insertions + 1)
  const d_blocks = Math.log10(deletions + 1)
  
  const i_part = (insertions <= 0) ? '' : `
    <span class="insertions n">+${insertions}</span>
    <span class="insertions-bg bar" style="width: ${i_blocks}em"></span> `
  const d_part = (deletions <= 0) ? '' : `
    <span class="deletions n">+${deletions}</span>
    <span class="deletions-bg bar" style="width: ${d_blocks}em"></span> `
  
  return `
    <span class="churn-bar">
      <span class="num-bar-container">${i_part}</span>
      <span class="num-bar-container">${d_part}</span>
    </span>`
}
