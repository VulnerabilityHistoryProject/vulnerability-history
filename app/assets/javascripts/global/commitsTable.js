/*
 Generic table to be used for commits
 Lists the commit no., commit message, and commit date
 Currently being used on filepaths/show, but should be reused as needed
 */

function buildTopNav() {
    $('#datatable_wrapper').prepend('<div id="table-nav-top"></div>');
    const nav = $('#table-nav-top');
    nav.addClass('table-nav grid-x');

    const nav_length = $('#datatable_length');
    nav_length.addClass('table-nav-show-entries cell small-9')
    nav.append(nav_length);

    const nav_search = $('#datatable_filter label');
    nav_search.addClass('table-nav-search cell small-3')
    nav.append(nav_search);
}

function languageInfo() {
    return {
        search: "", // nothing in the placeholder;
        searchPlaceholder: 'Search',
        lengthMenu: "Show _MENU_",
        info: "Showing _START_ to _END_ of _TOTAL_ commits",
        infoEmpty: "Showing 0 to 0 of 0 entries",
        infoFiltered: "(filtered from _MAX_ total commits)",
    };
}

function editsTable(jsonData, tableName = '#edit_table') {
    $(tableName).dataTable({
        responsive: true,
        destroy: true,
        data: jsonData,
        order: [[2, "desc"]], /* Default sort */
        columnDefs: [],
        columns: [
            {
                title: 'Commit No',
                data: 'commit_hash',
                defaultContent: '',
                className: 'title',
                responsivePriority: 1
            },
            {
                title: 'Commit Message',
                data: 'message',
                defaultContent: '',
                responsivePriority: 2
            },
            {
                title: 'Date',
                data: 'date_created',
                defaultContent: '',
                responsivePriority: 3
            }
            ],
        language: languageInfo(),
        initComplete: function () {
            $('#loading').remove();
            buildTopNav();
        }
    });

    $('#edit_table').on('mouseover', 'tbody tr', function () {
        $(this).css({
            'cursor': 'pointer',
            'color': 'blue'
        });
    })

    $('#edit_table').on('mouseout', 'tbody tr', function () {
        $(this).css({
            'cursor': 'default',
            'color': 'inherit'
        });
    });
};
