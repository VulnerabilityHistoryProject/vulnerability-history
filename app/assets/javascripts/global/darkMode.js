const toggleBtn = document.getElementById("themeToggle");

function prefersLightTheme(){
	return window.matchMedia("(prefers-color-scheme: light)").matches;
}

function setThemeDark(isDark){
	document.body.classList.add(isDark ? 'dark-mode' : 'light-mode');
	document.body.classList.remove(isDark ?  'light-mode' : 'dark-mode');
	toggleBtn.checked = isDark;
	localStorage.setItem("use-dark-mode", isDark)
}

function currentThemeIsDark(){
	return document.body.classList.contains("dark-mode")
}

function decideUseDark(){
	switch(localStorage.getItem("use-dark-mode")){
		case "true":
			return true
		case "false":
			return false
		default:
			return prefersLightTheme()
			// if you have zero preferences and history... DARK MODE!
			// ...because Andy likes VHP dark mode better.
	}
}

/*
  How we determine light/dark mode:

  Precedence:
	1. localStorage
	2. perfers-color-scheme
	3. Dark mode

  Then: store in localStorage after determine from precedence
 */
export default function determineLightDarkMode() {
	const useDarkMode = decideUseDark();
	toggleBtn.addEventListener("click",() => setThemeDark(!currentThemeIsDark()));
	setThemeDark(useDarkMode)
}