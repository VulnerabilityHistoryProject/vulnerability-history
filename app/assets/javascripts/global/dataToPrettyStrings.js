/*
This file is a small utility for formatting data coming from the VHP API
It's mostly just string concatentation with some minor logic.
So that we keep things consistent.
*/

export function vulnProject(data){
    return `<img class="${data.subdomain}-inline-logo inline-logo"
                 title="${data.project_name}"/>`
}

export function vulnName(data) {
  let str = data.cve;
  if (data.nickname != "" && data.nickname != null) {
    str = `<span class="nicknamed-cve"> ${data.nickname}</span>
           <span class="nicknamed-aka-cve"> aka ${data.cve}</span>`;
  }
  return str;
}

export function vulnLink(data, str) {
  return `<a data-id="${data.id}" class="dt_row" href="/${data.id}">${str}</a>`
}

//Simplifies the CWE name if it has a nickname in parentheses
export function friendlyCWE(name) {
  // "I do not want this ('I want this') not this" --> "I want this"
  // "Just a string" --> "Just a string"
  const match = name.match(/.*\(\'(.+)\'\).*/);
  if (match) {
    return match[1]; //[0] is the full string, [1] is the actual match
  } else {
    return name;
  }
}

export function ellipsizedFilepath(fp, prefixLen, maxLen) {
  /* Explanation of this magical regex
     Group 1: from the beginning of the string up to prefixLen, greedily get
              the biggest string you can ending in a /
              e.g. foo/bar/baz --> foo/ for prefixLen=3
     Group 2: at least one character must be here to be ellipsized/discarded,
              again note that we're greedy here so we ellipsize as much as
              possible before getting to group 3.
     Group 3: from the end of the string, get the smallest possible string
              starting with a / (i.e. filename)
  */
  const regex = new RegExp(`^(.{1,${prefixLen}}/)(.+)(/.+)$`, "g");
  const replacer = function (str, prefix, _mid, suffix) {
    if (str.length > maxLen) {
      return `${prefix}...${suffix}`;
    } else {
      return str;
    }
  };
  return [replacer, fp.replace(regex, replacer)];
}
