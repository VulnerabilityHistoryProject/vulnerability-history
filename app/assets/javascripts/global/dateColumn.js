function renderDateColumn(data, type, row, displayFormat, isLink) {
  if (type === 'display') {
    // Display dates in a nice format.
    if (isLink) {
      return renderAsLink(moment(data).format(displayFormat), row);
    } else {
      return moment(data).format(displayFormat);
    }
  } else if (type === 'filter') {
    // Allow for the filtering of dates in a variety of formats.
    // The type is "filter" when the search bar is used.
    let currentDate = moment(data);
    let longDisplay = currentDate.format(displayFormat);
    let shortDisplay = currentDate.format("MM/DD/YYYY");
    return data + " " + longDisplay + " " + shortDisplay;
  } else {
    // Sort dates without any special formatting.
    return data;
  }
}
