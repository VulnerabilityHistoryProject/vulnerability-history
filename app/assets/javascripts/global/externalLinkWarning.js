$(document).ready(function(){
    var root = new RegExp(location.host);

    $('a').each(function(){

        if(root.test($(this).attr('href'))){
            $(this).addClass('local');
        }
        else{
            // a link that does not contain the current host
            var url = $(this).attr('href');
            if(url) {
                if(url.length > 1)
                {
                    $(this).addClass('external');
                }
            }
        }
    });

    $('a.external').on('click', function(e){

        e.preventDefault();
        var answer = confirm("You are about to leave the Vulnerability History Project website and view the content of an external website. Do you want to proceed?");

        if (answer){
            window.location = $(this).attr('href');
        }

    });

});
