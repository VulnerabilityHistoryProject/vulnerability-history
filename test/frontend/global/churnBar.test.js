import {churnBar} from '../../../app/frontend/global/churnBar'


// SNAPSHOT
test("ChurnBar", () => {
    const test = churnBar(20, 80);
    expect(test).toMatchSnapshot();
});
