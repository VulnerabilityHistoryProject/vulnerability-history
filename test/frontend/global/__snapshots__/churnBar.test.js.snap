// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChurnBar 1`] = `
"
    <span class=\\"churn-bar\\">
      <span class=\\"num-bar-container\\">
    <span class=\\"insertions n\\">+20</span>
    <span class=\\"insertions-bg bar\\" style=\\"width: 1.3222192947339193em\\"></span> </span>
      <span class=\\"num-bar-container\\">
    <span class=\\"deletions n\\">+80</span>
    <span class=\\"deletions-bg bar\\" style=\\"width: 1.9084850188786497em\\"></span> </span>
    </span>"
`;
